### ISSUE TITLE
Insufficient permission to access a field after unfolding a predicate 

### ISSUE THREAD

**Issue by mario-bucev** [viperproject/silicon#481]
Hi,
Carbon successfully verifies the following code, but Silicon doesn't:
```silver
field val: Int

function read(): Perm
  ensures none < result
  ensures result < write

predicate P(self: Ref) {
  acc(self.val, write)
}

predicate array(arr: Seq[Ref]) {
  // If we replace the forall with acc(P(arr[1])), the verification succeeds
  |arr| == 3 && forall i: Int :: { arr[i] } 0 <= i && i < |arr| ==> acc(P(arr[i]))
}

method test() returns (res: Ref)
{
  var arr: Seq[Ref]
  inhale acc(array(arr), read())
  unfold acc(array(arr), read())
  unfold acc(P(arr[1]), read())

  inhale acc(res.val, write)
  // Assignment might fail. There might be insufficient permission to access res.val
  res.val := 1
}
```
The problem may be the same as [this issue](https://github.com/viperproject/silicon/issues/479), though it seems that in the present case, the `forall` is problematic.

**Comment by mschwerhoff** [comment]
Potentially related: #367, #479, #480

**Comment by mschwerhoff** [comment]
Resolved with commit 6df3598b7c2c92d09c15a41f187f7ba2d509f495.
